### PR TITLE
drivers: counter: Add user_data to alarm callback

### PR DESCRIPTION
**Pull request by nordic-krch**
Modify alarm callback to return user_data and channel_id.
Set_alarm and disable_alarm updated accordingly.

Updated counter.h

Nordic drivers and test aligned.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Comment by nordic-krch** [comment]
fyi: @MaureenHelm @fvincenzo @SavinayDharmappa @stanislav-poboril

File: include/counter.h [hunk 1]
```diff
@@ -33,12 +33,11 @@ struct counter_alarm_cfg;
...
+ * @param user_data User data.
+ * @param chan_id   Channel ID.
+ * @param ticks     Counter value that triggered the callback.
  */
-typedef void (*counter_alarm_callback_t)(struct device *dev,
-					 const struct counter_alarm_cfg *cfg,
+typedef void (*counter_alarm_callback_t)(void *user_data, u8_t chan_id,
 					 u32_t ticks);
 
 /** @brief Alarm callback structure.
```
Comment (mnkp) [inline include/counter.h]: The sentence "Beware such structure should not be allocated on stack." on line 46 should be removed. We could mention in the description of `counter_set_channel_alarm()` function that the `alarm_cfg` parameter does not require static memory allocation.

**Review by mnkp — COMMENTED** [review]
Looks good, with some minor comments. Also Shippable needs to be made happy.

File: include/counter.h [hunk 2]
```diff
@@ -60,9 +59,7 @@ struct counter_alarm_cfg {
 		      *   counter implementation may count asynchronous events.
 		      */
 
-	u8_t channel_id; /*!< Channel ID. Number of channels is driver
-			  * dependent.
-			  */
+	void *user_data; /*!< User data returned in callback.*/
```
Comment (mnkp) [inline include/counter.h]: This is somehow unrelated to this PR but still needs to be addressed. I've generated locally Zephyr's documentation with `ninja htmldocs` and as expected the `/*!< User data returned in callback.*/` placed after the field name was ignored by doxygen. A working example of documenting structure fields can be looked up among others in i2s.h API searching for `struct i2s_config`.

**Comment by nordic-krch** [comment]
Added renaming `counter_set_ch_alarm` to `counter_set_channel_alarm`. 

@anangl suggested offline that `dev` could also be removed from `counter_wrap_callback_t`. This way it will be consistent with alarm callback. However, there is a backward compatibility issue. Once device is removed from wrap callback there is no straightforward way for having backward compatibility (`counter_set_alarm`) so I'm not sure what to do.
If we favor more consistent API over backward compatibility, then we can add optional backward compatibility to the driver (kconfig option) which would provide `callback(dev, user_data)`. It can even be runtime per instance. What do you think? @nashif any practices you already took for cases like that?

File: include/counter.h [hunk 3]
```diff
@@ -75,8 +75,7 @@ struct counter_alarm_cfg {
 typedef void (*counter_wrap_callback_t)(struct device *dev, void *user_data);
 
 /* Deprecated counter callback. */
-typedef void (*counter_callback_t)(struct device *dev,
-						void *user_data);
+typedef void (*counter_callback_t)(struct device *dev, void *user_data);
```
Comment (MaureenHelm) [inline include/counter.h]: Why are you changing the deprecated callback in this commit? It seems like an unrelated change to the renaming of `counter_set_ch_alarm` to `counter_set_channel_alarm`

**Review by MaureenHelm — DISMISSED** [review]

File: include/counter.h [hunk 4]
```diff
@@ -77,7 +74,8 @@ struct counter_alarm_cfg {
  */
 typedef void (*counter_wrap_callback_t)(struct device *dev, void *user_data);
 
-__deprecated typedef void (*counter_callback_t)(struct device *dev,
+/* Deprecated counter callback. */
+typedef void (*counter_callback_t)(struct device *dev,
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/counter.h]: Unfortunately, there is compilation warning if __depracate is there even when none of deprecated elements is used. It is caused by deprecated `set_alarm` function using `counter_callback_t` as argument.

File: include/counter.h [hunk 5]
```diff
@@ -102,10 +100,9 @@ struct counter_config_info {
 typedef int (*counter_api_start)(struct device *dev);
 typedef int (*counter_api_stop)(struct device *dev);
 typedef u32_t (*counter_api_read)(struct device *dev);
-typedef int (*counter_api_set_alarm)(struct device *dev,
-				const struct counter_alarm_cfg *alarm_cfg);
-typedef int (*counter_api_disable_alarm)(struct device *dev,
+typedef int (*counter_api_set_alarm)(struct device *dev, u8_t chan_id,
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/counter.h]: then driver must implement some kind of allocator. In case of single instance watchdog which is shared between multiple subsystems that's good approach but in case of counter it will most likely be used by single subsystem or object so allocator, imo,  is unnecessary complication of the driver.

File: include/counter.h (hunk 3 above)
Comment (nordic-krch) [reply include/counter.h]: yep, i will clean up a bit this PR since it became aggregation of couple of changes. Will split them into separate PRs

**Comment by mnkp** [comment]
> @anangl suggested offline that `dev` could also be removed from `counter_wrap_callback_t`. This way it will be consistent with alarm callback. However, there is a backward compatibility issue. Once device is removed from wrap callback there is no straightforward way for having backward compatibility (`counter_set_alarm`) so I'm not sure what to do.
> If we favor more consistent API over backward compatibility, then we can add optional backward compatibility to the driver (kconfig option) which would provide `callback(dev, user_data)`. It can even be runtime per instance.

One extra vote for a consistent API. Since we seem to agree that passing `dev` and `user_data` in the same call is redundant let's remove `dev` from `counter_wrap_callback_t` too.

As far as I know there is no official 'best practices' document on how to do API changes, even though there were talks on creating one. The preferred way however is to use `__deprecated` attribute. We did once an SPI API change via a Kconfig option where user could select old or new API and that turned out to be a bit cumbersome, some people were not happy about it. One disadvantage of using `__deprecated` is a need to come up with new function names like `counter_set_channel_alarm` even though the old names like `counter_set_alarm` are much better.

In case of counter API we can use `__deprecated` and keep API backward compatible if we stay with
```
__deprecated static inline int counter_set_alarm(struct device *dev,
				    counter_callback_t callback,
				    u32_t count, void *user_data)
{
	const struct counter_driver_api *api = dev->driver_api;

	return api->set_alarm(dev, callback, count, user_data);
}
```
rather than doing
```
	return counter_set_wrap(dev, count, callback, user_data);
```
since that's where all the problems come from.

File: include/counter.h (hunk 2 above)
Comment (nordic-krch) [reply include/counter.h]: fixed

**Comment by nordic-krch** [comment]
@mnkp, issue is with legacy callback having `dev` and `user_data` and by removing `dev` from counter callbacks we have no option to pass both using simple wrapper. That's why i've proposed to extend driver implementation with backward compatibility extension. I've made a draft of that (https://github.com/nordic-krch/zephyr/commit/e6eafe4c9f46f0246ea4db3f0bc55a32843d83f3).

File: include/counter.h [hunk 6]
```diff
@@ -29,45 +29,35 @@
...
-
-	u8_t channel_id; /*!< Channel ID. Number of channels is driver
-			  * dependent.
-			  */
-
-	bool absolute; /*!< If true ticks are treated as absolute value, else
-			*   it is relative to the counter reading performed
-			*   during the call.
-			*/
+	counter_alarm_callback_t handler;
```
Comment (mnkp) [inline include/counter.h]: Let's rename this field to `callback`. That's also how other APIs e.g. watchdog.h name this field.

File: include/counter.h [hunk 7]
```diff
@@ -33,12 +33,11 @@ struct counter_alarm_cfg;
...
- * @param dev     Pointer to the device structure for the driver instance.
- * @param cfg     Original structure owning this handler.
- * @param ticks   Counter value that triggered the callback.
+ * @param user_data User data.
+ * @param chan_id   Channel ID.
+ * @param ticks     Counter value that triggered the callback.
  */
-typedef void (*counter_alarm_callback_t)(struct device *dev,
-					 const struct counter_alarm_cfg *cfg,
+typedef void (*counter_alarm_callback_t)(void *user_data, u8_t chan_id,
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply include/counter.h]: @MaureenHelm can you please summarize how to implement more advanced use cases like the one described in https://github.com/zephyrproject-rtos/zephyr/pull/8340#discussion_r237095779 without making this change?

[the remark this answers is not in the document]
Comment (pabigot) [reply include/counter.h]: See #11976 for rationale for providing both dev and a generic pointer.

[the remark this answers is not in the document]
Comment (anangl) [reply include/counter.h]: After reading #11976 and re-thinking this particular case, I think we should keep the `dev` pointer and add the generic one as the additional parameter. Apart from the semi-consistency argument raised by @MaureenHelm (see https://github.com/zephyrproject-rtos/zephyr/pull/8340#discussion_r236914015) I can imagine cases where using a common `user_data` for multiple devices would be desirable and then having only one generic pointer would force the users to use an additional structure for just holding both the `dev` and `user_data` pointers. This does not seem to be reasonable. Besides, the `dev` pointer "comes for free" (we'll almost always have it at hand in the routine that will call the callback, no need to use additional storage for it) so just passing it as the first parameter will not be a big cost.

[the remark this answers is not in the document]
Comment (mnkp) [reply include/counter.h]: > I can imagine cases where using a common `user_data` for multiple devices would be desirable and then having only one generic pointer would force the users to use an additional structure for just holding both the `dev` and `user_data` pointers.

The GPIO case described in #11976 is a bit different. Had GPIO callback passed a generic user_data pointer as a parameter there would be no problem getting access to all the required data.

> Besides, the `dev` pointer "comes for free"

It does come cheap. But - at least on ARM architecture - only up to 4 parameters are passed to the function as register variables. If the parameter list is longer we need to put them on the stack. So if we decide to have a general convention to always pass pointer to `dev` and to `user_data` there is a price to pay.

So far I haven't seen anywhere in the Zephyr's code (or an application) a place where having a pointer to `user_data` passed in the callback function would turn out to be insufficient. If someone could point to an example that proves to the contrary that certainly would be major argument for passing both parameters.

[the remark this answers is not in the document]
Comment (pabigot) [reply include/counter.h]: I don't think there's any argument that it isn't technically sufficient to pass only `user_data`.  It's just that in some cases that requires the user of the API to do extra work to capture the desired non-device pointer and the device pointer into an persistent object that wouldn't be needed if the handler were given both values.  It's not quite "no problem".

[the remark this answers is not in the document]
Comment (mnkp) [reply include/counter.h]: Sorry, by "would turn out to be insufficient" I actually meant

> requires the user of the API to do extra work to capture the desired non-device pointer and the device pointer into an persistent object that wouldn't be needed if the handler were given both values.

I really haven't seen it. Could you point me to a place in the code where such plumbing was necessary? In the GPIO case the callback doesn't pass `user_data` so the example is not adequate.

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+	zassert_true(user_data == exp_user_data, "Unexpected callback\n");
+	wrap_cnt++;
+}
+
+#include <nrf_gpio.h>
+void test_wrap_alarm_instance(const char *dev_name)
+{
+	struct device *dev;
+	int err;
+	int cnt;
```
Comment (SavinayDharmappa) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c]: @nordic-krch  isn't cnt variable should be u32_t ? In case 0f aonpt it is causing overflow as  aonpt is down counter and intial value of counter 0xffffffff.

Comment (nordic-krch) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c]: yes, i've updated that in this PR.

File: include/counter.h (hunk 6 above)
Comment (nordic-krch) [reply include/counter.h]: renamed.

**Comment by nordic-krch** [comment]
@mnkp @anangl @MaureenHelm I've added `struct device *dev` to alarm callback. Updated nordic implementations and test.

**Comment by anangl** [comment]
I think we should also rename the `counter_count_up` function because currently its name may suggest that it requests the counter to count up, not just checks the counting type.
What about `counter_counts_up` or `counter_is_counting_up`?
I am not sure if such renaming is worth a separate PR, maybe it could be done here?

**Comment by nordic-krch** [comment]
@anangl added that to this commit

**Comment by mnkp** [comment]
I have a few more comments regarding function names. I'm however also not sure if they should be handled in this PR.

We have `counter_set_channel_alarm` and its counterpart `counter_disable_channel_alarm`. Maybe we should rename the latter to `counter_clear_channel_alarm`. APIs typically use set/clear and enable/disable in pairs without mixing them.

Several functions use 'ticks' parameter to pass counter value. Ticks has an immediate connotation with kernel ticks and is therefore very confusing. IMHO the old 'count' name was much better. Also function names like `counter_ticks_to_us` will cause issues, especially if kernel will use counter API to provide actual kernel ticks.

We have a function `counter_set_wrap`. I'm not a native English speaker but the 'wrap' part somehow doesn't feel right. I've also never seen it used in the context of the counter. Not in an API, not in a datasheet. I understand it's a tricky one as our counter can count up or down so the 'reload_value' doesn't work. But what about `counter_set_top_value`, `counter_set_max_value`, `counter_set_compare_value`? These are names used in the datasheets and users are more familiar with their meaning.

**Comment by nordic-krch** [comment]
> We have counter_set_channel_alarm and its counterpart counter_disable_channel_alarm. Maybe we should rename the latter to counter_clear_channel_alarm.

`clear_channel_alarm` may be confusing as it suggests as if some kind of alarm flag got cleaned. Disable explains exactly what happens. If we look for symmetry then maybe `set_channel_alarm`-> `enable_channel_alarm` would be better?

>  Ticks has an immediate connotation with kernel ticks and is therefore very confusing.

That's subjective. I don't have such connotations. If we change parameter to count then we must have `counter_count_to_us`. That is confusing. 

> But what about counter_set_top_value, counter_set_max_value

I agree that `wrap` is not perfect. On the other hand, `set_wrap` sets periodic alarm and `set_top_value` or `set_max_value` does not sound like they are doing anything more than setting a value. What about something like `set/enable_main_alarm` (with complementary `channel_alarms`)?

**Comment by anangl** [comment]
> Disable explains exactly what happens. If we look for symmetry then maybe `set_channel_alarm`-> `enable_channel_alarm` would be better?

Yes, `enable/disable_channel_alarm` looks reasonable to me.

> I agree that `wrap` is not perfect. On the other hand, `set_wrap` sets periodic alarm and `set_top_value` or `set_max_value` does not sound like they are doing anything more than setting a value. What about something like `set/enable_main_alarm` (with complementary `channel_alarms`)?

Well, `set/enable_main_alarm` would suggest that the setting of the alarm is the only action taken. And the main purpose of this function is to set the wrap value, the callback to be associated with reaching this value is optional.

File: include/counter.h [hunk 9]
```diff
@@ -77,35 +69,29 @@ struct counter_alarm_cfg {
...
+	u8_t channels;
 };
 
 typedef int (*counter_api_start)(struct device *dev);
 typedef int (*counter_api_stop)(struct device *dev);
 typedef u32_t (*counter_api_read)(struct device *dev);
-typedef int (*counter_api_set_alarm)(struct device *dev,
-				const struct counter_alarm_cfg *alarm_cfg);
-typedef int (*counter_api_disable_alarm)(struct device *dev,
+typedef int (*counter_api_set_alarm)(struct device *dev, u8_t chan_id,
```
[the remark this answers is not in the document]
Comment (SavinayDharmappa) [reply include/counter.h]: @mnkp #9726  address the issue  of CI broken.

**Comment by mnkp** [comment]
> Yes, enable/disable_channel_alarm looks reasonable to me.

+1

I agree with comments from @anangl. The main purpose of the `counter_set_wrap` function is to set the reload value in the count down mode and the compare/top value in the count up mode. The possibility to trigger an alarm is more of a side effect.

I would like to insist on reconsidering usage of the 'ticks' name:
- In realm of RTOS 'tick' has a very specific meaning.
- In kernel a 'tick' happens when the system timer overflows, in our case 'tick' happens on every clock.
- Taking into account that Zephyr is targeted at low power IoT devices it's expected that a good share of users will want to configure kernel in tick-less mode and use a low power counter to supply kernel clock. The counter API usage of 'tick' will be very confusing.

That said I agree with the following
> If we change parameter to count then we must have counter_count_to_us. That is confusing.

I don't have a good alternative and having the function is definitely useful.

**Comment by anangl** [comment]
> I would like to insist on reconsidering usage of the 'ticks' name:
> 
> * In realm of RTOS 'tick' has a very specific meaning.
> * In kernel a 'tick' happens when the system timer overflows, in our case 'tick' happens on every clock.
> * Taking into account that Zephyr is targeted at low power IoT devices it's expected that a good share of users will want to configure kernel in tick-less mode and use a low power counter to supply kernel clock. The counter API usage of 'tick' will be very confusing.

I agree with @nordic-krch that this is a subjective thing if one connects ticks primarily with kernel. To me it feels natural to use the term tick also for counters and I personally don't find it confusing even in the context of an RTOS. But I don't really have a strong opinion about this.
@pizi-nordic You created #12068 and you'd want to use the counter API in the implementation of the system clock source that provides kernel ticks. What's your opinion in this topic, do you find this naming confusing?

**Comment by pizi-nordic** [comment]
@anangl, @mnkp: I had similar discussion with @andyross when he reworked timer subsystem of the Zephyr. By my understanding the nomenclature is following:

- Hardware timers/counters are counting in CYCLES.
- Kernel uses a different time unit, as on some architectures hardware counters are pretty fast and we would have to use 64-bit calculations to deal with reasonable timeouts (like 30s). This unit is named TICK and consist of integer number of CYCLES.
- The timer interrupt announces how many TICKS passed from last announcement. As this is done only when needed, we could say that Zephyr is tickless (there is no regular clock ticks, announcement is scheduled when kernel needs a wake up signal).

So as you see the TICK in Zephyr (time unit) is much closer to "jiffy" in Linux (also time unit) than classic
tick definition (regular timer interrupt). And this is a bit confusing.

**Comment by pabigot** [comment]
@nordic-krch this one also needs to be converted to a cherry-pick of the one commit.  Rebasing makes a mess.  I'm not sure how the disable of `power_test` got into this branch either (well, I know "how", I don't know "why").

No rush, now that I know what's going on I'll cherry-pick the pieces I need.

**Review by pabigot — APPROVED** [review]

**Comment by nordic-krch** [comment]
proposed change of naming as suggested by @mnkp  would look like:
`counter_set_wrap` -> `counter_set_top_value`
`counter_get_wrap` -> `counter_get_top_value`
`counter_get_max_wrap` -> `counter_get_max_top_value`

@mnkp @pabigot any counter-proposals?

**Comment by nordic-krch** [comment]
@pabigot @mnkp wrap stuff renamed, please re-review.

**Comment by pabigot** [comment]
@nordic-krch I think it would be best to move the top rename out of this PR.  Looking at it I'm noticing some things I think are problematic with the API and need to think about and describe, and I don't want to delay merging the callback rework since it's blocking other platforms from updating.

**Review by pabigot — APPROVED** [review]
Never mind waiting; of course this changes API that people have to update for too.

One non-blocking comment.

**Review by mnkp — APPROVED** [review]

**Comment by mnkp** [comment]
It's a bit late but this PR is missing one change that we agreed to perform. 
> We have counter_set_channel_alarm and its counterpart counter_disable_channel_alarm. Maybe we should rename the latter to counter_clear_channel_alarm. APIs typically use set/clear and enable/disable in pairs without mixing them.

> > Disable explains exactly what happens. If we look for symmetry then maybe `set_channel_alarm`-> `enable_channel_alarm` would be better?
> 
> Yes, `enable/disable_channel_alarm` looks reasonable to me.

It would be good to align these names before `topic-counters` is merged.

**Comment by nordic-krch** [comment]
@mnkp yes, forgot about it. Lets wait with that until all remaining PR's (3 now) will go into topic-counters. It will be easier to change that at once on all implementations then forcing yet another round of aligning for each PR.

**Comment by pabigot** [comment]
@mnkp @nordic-krch Completely agree about being consistent, but....

Somewhere in an issue comment it was clarified that `counter_set_channel_alarm` configures a one-shot alarm that is not automatically repeated when the counter wraps.  That's not yet in the documentation AFAICT, and maybe my recollection is wrong.

In my opinion if this is a one-shot alarm it should be set/cancel.  If it's a repeating alarm it should be enable/disable.

Does this work for everybody?
